### PR TITLE
Fix retry issue for closed client requests

### PR DESCRIPTION
**Pull request by jeet1995**
This PR bubbles up `InternalServerErrorException`, an instance of `CosmosException` when unhandled / unwrapped errors are thrown from the `RntbdTransportClient` with status code 500 `Closed Client` (500:20912) to maintain consistency of `CosmosException` thrown for read and write operations. One example of an unwrapped error is the `ClosedClientTransportException` - a new sub-type of `TransportException` which forms the inner exception when operations are executed for a `CosmosClient` and `CosmosAsyncClient` instance that is closed. Previously for writes - a `TransportException` would be thrown to the caller in the downstream application.

This PR can also be considered as a bug fix in response to a customer reported incident where the SDK performed unnecessary 404 `Read Session Not Available` retries because `InternalServerErrorException` was not bubbled up correctly in case of operations with read semantics using `Session` consistency and executed through a closed `CosmosClient` and `CosmosAsyncClient` instance - see #41152

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java [hunk 1]
```diff
@@ -211,56 +212,70 @@ Mono<StoreResponse> writePrivateAsync(
...
 
+                    if (cosmosException != null) {
+                        return Mono.error(cosmosException);
+                    }
+
+                    String errorMessage = "Unexpected exception " + unwrappedException.getMessage() + " received while reading from store.";
+                    return Mono.error(new InternalServerErrorException(
+                        com.azure.cosmos.implementation.Exceptions.getInternalServerErrorMessage(errorMessage),
+                        unwrappedException,
+                        HttpConstants.SubStatusCodes.INVALID_RESULT));
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java:277]: should we use a different substrates code?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java [hunk 2]
```diff
@@ -420,7 +421,7 @@ private static Tag tag(long id) {
 
     private void throwIfClosed() {
         if (this.closed.get()) {
-            throw new TransportException(lenientFormat("%s is closed", this), null);
+            throw new ClosedClientTransportException(lenientFormat("%s is closed", this), null);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:424]: Curious: why not directly throw a InternalServerError here?

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java (hunk 2 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:424]: My reasoning is this - we do a wrapping of `InternalServerErrorException` in the outer layers such as `StoreReader` - so wanted to throw `ClosedClientTransportException` instead to keep the changes isolated. Also, I make a decision to re-throw (in `ConsistencyReader`) if I know the first-level inner type is `ClosedClientTransportException`.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java (hunk 1 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java:277]: Good idea!

Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java]: Fixed in next iteration.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java (hunk 2 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java]: Revisiting this - throwing `InternalServerErrorException` from `RntbdTransportClient` directly.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java [hunk 3]
```diff
@@ -211,56 +213,64 @@ Mono<StoreResponse> writePrivateAsync(
...
+                                        }
+                                    } catch (Throwable throwable) {
+                                        if (throwable instanceof Error) {
+                                            logger.error("Unexpected failure in handling orig [{}]", t.getMessage(), t);
+                                            logger.error("Unexpected failure in handling orig [{}] : new [{}]", t.getMessage(), throwable.getMessage(), throwable);
+                                            throw (Error) throwable;
+                                        } else {
+                                            // this happens before any retry policy - like for example GoneAndRetryRetryPolicy
+                                            // kicks in - no need to spam warn/error level logs yet
+                                            logger.info("Unexpected failure in handling orig [{}]", t.getMessage(), t);
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java:254]: Why both - isn't this guaranteed to always be a duplicate?

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java (hunk 3 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java]: Fixed in next iteration.

**Pull request merged**